amd/src/login.js

```diff
@@ -0,0 +1,214 @@
+            // Populate the username field in the password form
+            $options.password.username.val(username);
+            $options.password.form.submit();
+        }
+    }
+
+    function handlePasswordSubmission($event) {
+        $event.preventDefault();
+        var $options = $event.data;
+        var username = $options.password.username.val();
+
+        queryWebservice(username)[0].done(function(response) {
+            var currentURL = new URL(M.cfg.wwwroot).hostname;
+            if(response.domain !== currentURL) {
+                handleDomainRedirect(response, username);
+            } else {
+                $options.password.form.submit();
+            }
+        });
+    }
+
+    function setRememberme($options) {
+        var $remFromUserForm = $options.username.rememberme;
+        var $remFromPassForm = $options.password.rememberme;
+        if ($remFromUserForm.prop('checked')) {
+            $remFromPassForm.val("1");
+        }
+    }
+
+    function setReturnurl($options) {
+        var retFromUserFormVal = $options.username.returnurl.val();
+        var $retFromPassForm = $options.password.returnurl;
+        if (retFromUserFormVal) {
+            $retFromPassForm.val(retFromUserFormVal);
+        }
+    }
+
+    function checkDomain(username, $options) {
+        var result = false;
+        queryWebservice(username)[0].done(function(response) {
+            result = handleDomainRedirect(response, username, $options);
+        }).fail(function() {
+            get_string_and_notify('danger', 'invalid_user');
+            result = false;
+        });
+        return result;
+    }
+
+    function queryWebservice(username){
+        return ajax.call([
+            {
+                methodname: WEB_SERVICE_METHOD_NAME,
+                args: {
+                    username: username
+                }
+            }
+        ]);
+    }
+
+    function handleDomainRedirect(response, username, $options) {
+        var currentURL = new URL(M.cfg.wwwroot);
+        if (response.email === null) {
+            // flag non-existent user error
+            get_string_and_notify('warning', 'non_existent_user');
+            return false;
+        } else if (response.domain !== currentURL.hostname) {
+            // Redirect to recorded user default domain
+            window.location = window.location.protocol + '//' + response.domain + currentURL.pathname +
+                '/local/signin/index.php?username=' + username;
+            return false;
+        } else {
+            $(defaults.form.alert).alert('close');
+            // Stay on page & flip over to password form
+            // Populate username field on password form
+            $options.password.username.val(username);
+            // Toggle the password form
+            $options.password.changeuser.click();
+            $options.password.submit.on('click', $options.dom, $options.password.validation);
+        }
+        return true;
+    }
```

Comment:
Two things:

* The word `handle` is a little misleading here -- this isn't an event handler. Could we go with `maybeRedirect()` or similar? The function is ensuring the user exists, testing the returned domain against the current `window.location` and then possibly initating a redirect.
* Why are we passing `$options` here? It should be within scope already.
* This function both has (unused, as far as I can see) return values and has side effects on the environment. It might be worth breaking it down further into one function that ensures the user exists, another that builds the URL to redirect to and a separate function to do the redirect.

Reply:
Options doesn't seem to be in scope within the function automatically. We have taken it out and the functionality breaks.